### ISSUE TITLE
[iOS]: trouble compiling / "generic parameter 'ElementOfResult' could not be inferred"

### ISSUE THREAD

**Issue by monsieurtanuki** [juliansteenbakker/mobile_scanner#544]
```
❌  /Users/runner/hostedtoolcache/flutter/stable-3.7.7-x64/.pub-cache/hosted/pub.dev/mobile_scanner-3.1.0/ios/Classes/SwiftMobileScannerPlugin.swift:41:43: generic parameter 'ElementOfResult' could not be inferred

                let barcodesMap = barcodes!.compactMap { barcode in
            ~~~ ^
```

**Comment by orzogc** [comment]
https://github.com/juliansteenbakker/mobile_scanner/issues/489#issuecomment-1432315480

**Comment by monsieurtanuki** [comment]
Thank you @orzogc for your comment!
That may help the project.
That won't help me (cf. https://github.com/juliansteenbakker/mobile_scanner/issues/489#issuecomment-1477687815)

**Comment by juliansteenbakker** [comment]
I don't have access to an older version of xcode, so im not sure how i can solve this issue. Everything builds fine on my end.

**Comment by M123-dev** [comment]
Seems like a somewhat widespread issue #489 #495, but at the same time the compactMap method seems like a very [straight forward concept](https://developer.apple.com/documentation/swift/sequence/compactmap(_:)) to me and it's used just twice in this repo. 

Why not replace it with the normal mapping method and iterate over it to remove the nils or create a type extension which does that.

 (Keep in mind I never directly worked in swift so these are just assumptions)

**Comment by juliansteenbakker** [comment]
I have replaced the compactMap function with a traditional for loop. You can try it with this branch: https://github.com/juliansteenbakker/mobile_scanner/tree/bug/map-ios

**Comment by monsieurtanuki** [comment]
@juliansteenbakker Some updates:
* it looks like using `var barcodesData: [Any?] = []` makes the compiler happy (not sure how relevant it is, though)
* even better, you can keep the `compactMap` syntax by just specifying explicitly the type: `let barcodesMap: [Any?] = barcodes!.compactMap {`
* it looks like `available(iOS 15.4, *)` cannot really work on an old mac, as `device.automaticallyAdjustsFaceDrivenAutoFocusEnabled` only exists in 15.4+. That means that at compile time on the computer, the field must exist. Then at runtime the code will be called or not, depending on the runtime version. As far as I understand.

The most problematic part is obviously the iOS 15.4 using a new field. Looks like a deal-breaker for "old" macbooks.

**Comment by monsieurtanuki** [comment]
Solved by #568.
